designs/0032-cmdstan-output-formats.md

```diff
@@ -0,0 +1,230 @@
+- Feature Name: cmdstan-output-formats
+- Start Date: 2022-01-11
+- RFC PR:
+- Stan Issue:
+
+# Summary
+[summary]: #summary
+
+Provide alternatives to the [Stan CSV file format](https://mc-stan.org/docs/cmdstan-guide/stan_csv.html)
+for outputs of the Stan platform services methods supported by CmdStan.
+
+# Motivation
+[motivation]: #motivation
+
+The [Stan CSV file format](https://mc-stan.org/docs/cmdstan-guide/stan_csv.html) is limited and limiting for several reasons.
+
+The CSV format itself is not fully standardized.
+The original intent of the format is to represent a table's worth of data in plain text.
+Each row of the file contains one row of table data, and all rows have the same number of fields.
+Over time, conventions have been adopted, notably use of row headers for column labels and comment rows.
+
+The Stan CSV file, however, contains far more information than a single table's worth of data.
+To cram all the information from a CmdStan run we 
+abuse the CSV format by using blocks of comment lines to record additional information.
+The current Stan API requires the information in a comment line be assembled and flattened into a text string.
+Even when we can use a fast CSV parser to extract the CSV data, we still need to build ad-hoc parsers for the comment blocks
```

Comment:
I would add a note here about the relative inefficiency of CSV compared to a binary format, and the lack of types (e.g. int/real distinctions).

On the other hand, it's worth noting that this plain-text format has the advantage of being very 'accessible' in the sense that no library support is really needed, just a lot of post-processing due to our additional information we add in.

Reply:
The original intent was just to have the matrix of draws data to read in pure CSV format (with comments, but not comments with content to read).